### PR TITLE
Carson scales free

### PR DESCRIPTION
**Pull request by cpsievert**
Howdy @chriddyp @mkcor @tdhock!

The purpose of this pull request is to fully implement `facet_wrap(..., scales = 'free')`.

It also seems possible to implement space = "free", but perhaps that should be in a separate pull request.

### CODE REVIEW

**Comment by cpsievert** [comment]
You might find [this script](https://gist.github.com/cpsievert/211357490a854b814613) I wrote helpful for comparing the output of `gg2list()` between two different packages. When I `Rscript json-diffs.R`, I can see colorized diffs in my console like this:

screen shot 2015-02-17 at 6 26 14 pm

Installing from GitHub is perhaps inefficient, but it might be worth avoiding nasty `git checkout` errors. Let me know if any ideas on how to improve it!

**Comment by cpsievert** [comment]
Ah, crud. I messed with the white space a bit which makes the last commit diff hard to read. You can add `?w=1` to the end of url to ignore whitespace, [for example, this is better to look at](https://github.com/cpsievert/plotly/commit/12f471a3e642eb112cfacbfcd988c2cab33a0a0f?w=1)

**Comment by mkcor** [comment]
@cpsievert Regarding whitespaces, I go for RStudio's indentation.  In this respect, it looks like some of your whitespace changes are 'fixes', so that's all good.  I'm not sure about others (see inline comments).

File: R/ggplotly.R [hunk 1]
```diff
@@ -30,19 +30,19 @@ polygon.line.defaults$colour <- NA
...
+  toRGB(col)
+},
+size=identity,
+sizeref=identity,
+sizemode=identity,
+alpha=identity,
+shape=function(pch) {
+  pch2symbol[as.character(pch)]
+},
+direction=identity)
```
Comment (mkcor) [inline R/ggplotly.R]: Why this change?  Please keep it the way RStudio does indentation.

File: R/ggplotly.R [hunk 2]
```diff
@@ -175,7 +178,7 @@ gg2list <- function(p){
     df <- merge(df, gglayout[, c("PANEL", "plotly.row", "COL")])
     df <- df[order(df$order),]
     df$order <- NULL
-
+
```
Comment (mkcor) [inline R/ggplotly.R]: For instance, this sounds good, i.e., coming from RStudio's editor.

File: man/ensure_file_exist.Rd [hunk 3]
```diff
@@ -0,0 +1,15 @@
+% Generated by roxygen2 (4.1.0): do not edit by hand
```
Comment (mkcor) [inline man/ensure_file_exist.Rd]: Agreed, this file should have been tracked.  Thank you.
And thank you for rerunning `roxygenize()`.

**Comment by cpsievert** [comment]
Sweet. That's what I prefer to.

I also prefer 

```r
f <- function() {
  print("hi")
}
```

over 

```r
f <- function() 
  {
    print("hi")
  }
```

so I'd be happy to change that as well

File: R/ggplotly.R (hunk 1 above)
Comment (cpsievert) [inline R/ggplotly.R]: Hmm, no good reason. I'll change that back

**Comment by chriddyp** [comment]
Nice! hey @tdhock is it possible/easy to run that test table on this branch?

**Comment by cpsievert** [comment]
I'm pretty sure Toby's current implementation would only support visual testing of one branch at a time. Perhaps we could do something similar in a "username.github.com" repo with a folder for each branch?

**Comment by tdhock** [comment]
yes it should be possible to run the test table on this branch, but I will need to do a little hacking. Will post a link to the updated table when I get a chance to work on it.

**Comment by tdhock** [comment]
check out the test table

http://ropensci.github.io/plotly-test-table/tables/314a6ec37d1d423cc90120903b0a1deb0b4eb717/index.html

there are some cases with weird disappearing boxes

http://ropensci.github.io/plotly-test-table/tables/314a6ec37d1d423cc90120903b0a1deb0b4eb717/ticks-boxes-scales-free.html

(not your fault, this test only exists on the toby-fixes branch)

https://github.com/ropensci/plotly/blob/toby-fixes/tests/testthat/test-ggplot-ticks.R#L51

**Comment by cpsievert** [comment]
Sweet, thanks @tdhock! Will toby-fixes be merged into master soonish? Or maybe we should merge carson-scales-free and toby-fixes?

**Comment by chriddyp** [comment]
Cool - @tdhock @mkcor - should we move these *.r files that have the cookbook examples branch (https://github.com/ropensci/plotly/tree/add-r-cookbook-tests/tests/cookbook-test-suite) into the testhat directory (or elsewhere) in master so that the test-table can pick up on that rich set of examples?
@tdhock I'm assuming that checkout out the `add-r-cookbook` branch and running those additional tests would add too much complexity to your table generating script (https://github.com/ropensci/plotly-test-table/blob/gh-pages/index.R)?
i'd love to just visually inspect all of the examples in the facets chapter (https://github.com/ropensci/plotly/blob/add-r-cookbook-tests/tests/cookbook-test-suite/facets.r) before merging this PR

**Comment by tdhock** [comment]
@cpsievert I don't know how long it will be before toby-fixes it merged into master, so I would suggest to pull changes from toby-fixes into your branch.

@chriddyp @mkcor I agree that it is a good idea to add examples from the cookbook branch to the test table. In my opinion the easiest way to do this would be to just add them as tests/testthat/test-*.R files under toby-fixes. I can do that if you guys tell me which of

https://github.com/ropensci/plotly/tree/add-r-cookbook-tests/tests/cookbook-test-suite

you would like to include... or would you like to include all of the *.r and *.R files?

**Comment by chriddyp** [comment]
@tdhock OK, that sounds good. All of them are good except `multiple_graphs_on_one_page.r`

**Comment by mkcor** [comment]
@cpsievert [On curly braces] Absolutely. It's not only a preference, it's our style guide (https://github.com/ropensci/plotly/wiki/Development-guidelines#coding-style, https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml#curlybraces).

**Comment by mkcor** [comment]
@tdhock @chriddyp Definitely, let's include the cookbook examples as regular `.R` files (not `.r`) under `tests/testthat/`.  But please create a new branch for this and keep each PR as small as possible!!

**Comment by mkcor** [comment]
@cpsievert @tdhock `toby-fixes` could be merged sooner if it was smaller. The bigger the PR, the longer it takes to review. It should not be `toby-fixes` in the first place: There should be `toby-errorbar-horizontal`, `toby-box-scales-free`, etc. Each branch and corresponding PR should address *one* issue.

**Comment by mkcor** [comment]
@cpsievert @tdhock For reference, https://github.com/ropensci/plotly/wiki/Development-guidelines#a-few-guidelines -- Thanks @chriddyp for updating and cleaning that up!

**Comment by mkcor** [comment]
@cpsievert Can you please revert c542520 ? You shouldn't find yourself merging another feature/dev/wip branch into your own feature branch. Merging `master` makes sense if it has changed meanwhile (basically, you would be merging `toby-fixes` indirectly, because it would have been merged into `master`, but this is not the case).
Once you revert this merge, I can do a final review.

**Comment by cpsievert** [comment]
@mkcor ok, I removed that merge from the history on this branch

**Comment by mkcor** [comment]
@cpsievert Amazing! Just pulled your branch for a quick final review.

File: R/ggplotly.R [hunk 4]
```diff
@@ -168,6 +170,9 @@ gg2list <- function(p){
     gglayout <- built$panel$layout
     ## invert rows so that plotly and ggplot2 show panels in the same order
     gglayout$plotly.row <- max(gglayout$ROW) - gglayout$ROW + 1
+    ## ugh, ggplot counts panel right-to-left & top-to-bottom
+    ## plotly count them right-to-left & *bottom-to-top*
```
Comment (mkcor) [inline R/ggplotly.R]: Please use a single `#` for comments (see https://github.com/ropensci/plotly/wiki/Development-guidelines#coding-style and https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml#comments).

File: tests/testthat/test-facets.R [hunk 5]
```diff
@@ -47,3 +47,49 @@ test_that("3 facets becomes 3 panels", {
...
+  free_y <- no_panels + facet_grid(vs~am, scales = "free_y")
+  info <- gg2list(free_y)
+  expect_axes(info, 1L)
+  expect_axes(info, 2L, "y")
+  
+  free_x <- no_panels + facet_grid(vs~am, scales = "free_x")
+  info <- gg2list(free_x)
+  expect_axes(info, 2L)
+  expect_axes(info, 1L, "y")
+})
```
Comment (mkcor) [inline tests/testthat/test-facets.R]: Please add the end of line.  Vim does it for you.  In RStudio you need to add a line break.

File: tests/testthat/test-facets.R [hunk 6]
```diff
@@ -47,3 +47,49 @@ test_that("3 facets becomes 3 panels", {
   u <- unique(trace.axes.df)
   expect_identical(nrow(u), 3L)
 })
+
+# expect a certain number of _unique_ [x/y] axes
+expect_axes <- function(info, n, axis = "x") {
+  pattern <- paste0("^", axis, "axis([0-9]+)?$")
+  #n.axes <- length(grep(pattern, names(info$kwargs$layout)))
```
Comment (mkcor) [inline tests/testthat/test-facets.R]: Do we need to keep this?

File: tests/testthat/test-facets.R [hunk 7]
```diff
@@ -47,3 +47,49 @@ test_that("3 facets becomes 3 panels", {
   u <- unique(trace.axes.df)
   expect_identical(nrow(u), 3L)
 })
+
+# expect a certain number of _unique_ [x/y] axes
```
Comment (mkcor) [inline tests/testthat/test-facets.R]: Great! This PR might be a good opportunity to rename `test-facets.R` into `test-ggplot-facets.R`.

File: tests/testthat/test-facets.R [hunk 8]
```diff
@@ -47,3 +47,49 @@ test_that("3 facets becomes 3 panels", {
...
+  axes <- with(info$kwargs, layout[grepl(pattern, names(layout))])
+  n.axes <- length(axes)
+  ranges <- do.call("rbind", lapply(axes, function(x) x$range))
+  expect_identical(nrow(unique(ranges)), as.integer(n))
+}
+
+no_panels <- ggplot(mtcars, aes(mpg, wt)) + geom_point()
+
+test_that("facet_wrap(..., scales = 'free') creates interior scales", {
+  free_both <- no_panels + facet_wrap(~am+vs, scales = "free")
```
Comment (mkcor) [inline tests/testthat/test-facets.R]: We want each of these new plots to be a part of the visual, image diff test suite.
Please add `save_outputs(free_both, "both_scales_free")`.

File: tests/testthat/test-facets.R [hunk 9]
```diff
@@ -47,3 +47,49 @@ test_that("3 facets becomes 3 panels", {
...
+  info <- gg2list(free_both)
+  expect_axes(info, 4L)
+  expect_axes(info, 4L, "y")
+  
+  free_y <- no_panels + facet_wrap(~am+vs, scales = "free_y")
```
Comment (mkcor) [inline tests/testthat/test-facets.R]: Please add `save_outputs(free_y, "y_scale_free")`, and so on.
So we can see these plots in @tdhock 's table!

File: tests/testthat/test-facets.R (hunk 6 above)
Comment (cpsievert) [inline tests/testthat/test-facets.R]: Nope.

File: tests/testthat/test-facets.R (hunk 7 above)
Comment (cpsievert) [inline tests/testthat/test-facets.R]: Is this renaming just to reflect that facets should behave exactly like they do in ggplot2? Just fyi, we could still implement `space = "free"`

Comment (mkcor) [inline tests/testthat/test-facets.R]: The file renaming is just to follow some convention: All test files that test for `ggplotly` conversions are named `test-ggplot-*.R`. Notice that other test files (such as `test-unimplemented.R` or `test-plotly-getfigure.R`) test for features other than ggplot --> plotly conversions.

**Comment by tdhock** [comment]
just re-made the test table on your branch

http://ropensci.github.io/plotly-test-table/tables/e1b76fffdaec64085aa2665a3b469a943014b8ee/index.html

I noticed a couple of regressions, can you please check them out?

http://ropensci.github.io/plotly-test-table/tables/e1b76fffdaec64085aa2665a3b469a943014b8ee/ticks-boxes-scales-free.html

http://ropensci.github.io/plotly-test-table/tables/e1b76fffdaec64085aa2665a3b469a943014b8ee/facet_wrap_free_y.html

http://ropensci.github.io/plotly-test-table/tables/e1b76fffdaec64085aa2665a3b469a943014b8ee/facet_wrap_free_x.html

http://ropensci.github.io/plotly-test-table/tables/e1b76fffdaec64085aa2665a3b469a943014b8ee/facet_grid_free.html

**Comment by tdhock** [comment]
by the way do you want to get added to ropensci so you can add a branch under the main plotly repos next time? i guess it doesn't really matter but I figured I would ask.

**Comment by cpsievert** [comment]
Thanks @tdhock. I'll hopefully have a fix soon. It'd be nice to be added, but not essential.

**Comment by chriddyp** [comment]
Hey @cpsievert - how is this going? anything I can do to help get this through? A few users wrote in about scales-free this weekend!

**Comment by cpsievert** [comment]
I just updated the test table (note that it didn't update automatically because this PR was initiated from my fork).

<http://ropensci.github.io/plotly-test-table/> 

I don't see any problems, and I'd be OK with merging now, but we can certainly add more of the cookbook tests if you'd like.

**Comment by chriddyp** [comment]
Fantastic, the visual tests look good! :+1: from me

**Comment by cpsievert** [comment]
Cool, thanks. Should I wait for a +1 from @mkcor?

**Comment by chriddyp** [comment]
@mkcor is on vacation till the end of the week, I think she'll be reviewing post-merge post-vacation. @tdhock could you take a closer look at the code for approval?

**Comment by cpsievert** [comment]
OK, I actually just noticed a regression in facet strip labels. I'll fix later today and let everyone know when it's ready.

**Comment by cpsievert** [comment]
Annotations are fixed now. Ready to merge when you are @chriddyp @tdhock

**Comment by cpsievert** [comment]
@tdhock it appears your latest push to plotly-test-table somehow overwrote homepage links to commits I made earlier today --

<http://ropensci.github.io/plotly-test-table/>

You can still see them here though --

<http://ropensci.github.io/plotly-test-table/tables/f800c1249fc36072c1b4dd775b88659144b80f17/>

**Comment by tdhock** [comment]
The homepage links are generated via the files in tables/* so I think the links will show up the next time the table is re-made.

**Comment by cpsievert** [comment]
@tdhock @chriddyp please have at the table and +1 (it looks good to me and I'm ready to merge)

http://ropensci.github.io/plotly-test-table/tables/f9e252af61276fd40168fc8916024877c7f2f6a3/

There was a failure in downloading contours (this seems to be quite common when I `make` locally):

http://ropensci.github.io/plotly-test-table/tables/f9e252af61276fd40168fc8916024877c7f2f6a3/contour.html

However, this test for the previous commit looks good (the only is difference is NEWS/DESCRIPTION)

http://ropensci.github.io/plotly-test-table/tables/f800c1249fc36072c1b4dd775b88659144b80f17/contour.html

**Comment by tdhock** [comment]
are you sure it is a download error? If so there should be a log file...

**Comment by cpsievert** [comment]
I don't see a log -- here is what I see in my terminal

screen shot 2015-03-10 at 2 18 14 pm

**Comment by tdhock** [comment]
If it is a download error you get a log line via this code https://github.com/ropensci/plotly-test-table/blob/gh-pages/table.R#L248-L250

but you are seeing an error due to this code https://github.com/ropensci/plotly-test-table/blob/gh-pages/table.R#L265-L270

the only reason why I can think that would be happening is that the py$ggplotly function returns an error
https://github.com/ropensci/plotly-test-table/blob/gh-pages/table.R#L233

that is the code for SENDING the plotly, not for downloading it. I guess I may need to add some code to attempt retries for the sending step as well.

**Comment by cpsievert** [comment]
OK, good to know, it's probably because I'm on a crappy wifi connection
right now. I don't think it should prevent this from being merged, agreed?

**Comment by cpsievert** [comment]
I made the table again (after merging with master which now has the fix for ribbon transparency) and everything looks good (with no caveats :)

http://ropensci.github.io/plotly-test-table/tables/bfa5bfef02ce059712ecb08d38106ddace3cd12e/index.html

Feel free to merge @chriddyp @tdhock :+1:

**Comment by tdhock** [comment]
+1
